### PR TITLE
Added duplicate finding functionality to ImageCleaner

### PR DESCRIPTION
**Pull request by fpingham**
With this new tool, users can easily and interactively delete duplicates from their datasets (both custom datasets and imported ones).

### CODE REVIEW

File: fastai/widgets/image_cleaner.py [hunk 1]
```diff
@@ -23,14 +25,76 @@ def padded_ds(ll_input, size=(250, 300), do_crop=False, padding_mode='zeros'):
...
+
+    def largest_indices(arr, n):
+        """Returns the n largest indices from a numpy array."""
+        flat = arr.flatten()
+        indices = np.argpartition(flat, -n)[-n:]
+        indices = indices[np.argsort(-flat[indices])]
+        return np.unravel_index(indices, arr.shape)
+
+    @classmethod
+    def sort_ids(cls, similarities):
```
Comment (sgugger) [inline fastai/widgets/image_cleaner.py]: sort_idxs, no?

Comment (fpingham) [reply fastai/widgets/image_cleaner.py]: Would be a better name, yes.

**Comment by sgugger** [comment]
Closed as discussed

File: fastai/widgets/image_cleaner.py [hunk 2]
```diff
@@ -23,14 +25,77 @@ def padded_ds(ll_input, size=(250, 300), do_crop=False, padding_mode='zeros'):
         "For a LabelList `ll_input`, resize each image in `ll_input` to size `size` by optional cropping (`do_crop`) or padding with `padding_mode`."
         return ll_input.transform(crop_pad(), size=size, do_crop=do_crop, padding_mode=padding_mode)
 
+    @classmethod
+    def from_similars(cls, learn, weight_file, layer_ls:list=[0, 7, 2], ds_type=DatasetType.Valid, **kwargs):
+        "Gets the indices for the most similar images in a dataset"
+        hook = hook_output(learn.model[layer_ls[0]][layer_ls[1]][layer_ls[2]])
+        if ds_type == DatasetType.Train: dl = DataLoader(learn.data.train_ds,
```
[the remark this answers is not in the document]
Comment (fpingham) [reply fastai/widgets/image_cleaner.py]: Please let me know once this is changed in fastai

File: fastai/widgets/image_cleaner.py [hunk 3]
```diff
@@ -23,14 +25,77 @@ def padded_ds(ll_input, size=(250, 300), do_crop=False, padding_mode='zeros'):
...
+    def comb_similarity(t1: torch.Tensor, t2: torch.Tensor, sim_func=nn.CosineSimilarity(dim=0)):
+        "Computes the similarity function `sim_func` between each embedding of `t1` and `t2` matrices. t1` and `t2` should have dimensions [n_embeddings, n_features]"
+        self_sim = False
+        if torch.equal(t1, t2): self_sim = True
+        print('Computing similarities...')
+
+        sims = np.zeros((t1.shape[0], t2.shape[0]))
+        for idx1 in progress_bar(range(t1.shape[0])):
+            for idx2 in range(t2.shape[0]):
+                if not self_sim or idx1>idx2:
```
[the remark this answers is not in the document]
Comment (fpingham) [reply fastai/widgets/image_cleaner.py]: Will do

File: fastai/widgets/image_cleaner.py [hunk 4]
```diff
@@ -60,7 +125,26 @@ def make_dropdown_widget(cls, description='Description', options=['Label 1', 'La
     def make_horizontal_box(cls, children, layout=Layout()): return widgets.HBox(children, layout=layout)
 
     @classmethod
-    def make_vertical_box(cls, children, layout=Layout()): return widgets.VBox(children, layout=layout)
+    def make_vertical_box(cls, children, layout=Layout(), duplicates=False):
+        if not duplicates: return widgets.VBox(children, layout=layout)
+        else: return widgets.VBox([children[0], children[2]], layout=layout)
+
+    def chunks(self, l, n):
```
[the remark this answers is not in the document]
Comment (sgugger) [reply fastai/widgets/image_cleaner.py]: The chunks function, I think you can move it in core.

File: fastai/widgets/image_cleaner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply fastai/widgets/image_cleaner.py]: Done. You can do dl = learn.data.train_dl.new(shuffle=False).
